app/src/main/java/org/kiwix/kiwixmobile/local_file_transfer/WifiDirectManager.kt

```diff
@@ -0,0 +1,395 @@
+        receiverDeviceAsyncTask = ReceiverDeviceAsyncTask(this).also {
+          it.execute()
+        }
+      }
+    }
+  }
+
+  fun changeStatus(itemIndex: Int, @FileStatus status: Int) {
+    filesForTransfer[itemIndex].fileStatus = status
+    callbacks.onFileStatusChanged(itemIndex)
+    if (status == FileStatus.ERROR) {
+      displayToast(
+        R.string.error_transferring, filesForTransfer[itemIndex].fileName,
+        Toast.LENGTH_SHORT
+      )
+    }
+  }
+
+  private fun cancelAsyncTasks(vararg tasks: AsyncTask<*, *, *>?) =
+    tasks.all {
+      it?.cancel(true) ?: false
+    }
+
+  fun stopWifiDirectManager() {
+    cancelAsyncTasks(peerGroupHandshakeAsyncTask, senderDeviceAsyncTask, receiverDeviceAsyncTask)
+    if (isFileSender) {
+      closeChannel()
+    } else {
+      disconnect()
+    }
+    unregisterWifiDirectBroadcastReceiver()
+  }
+
+  private fun disconnect() {
+    manager.removeGroup(channel, object : ActionListener {
+      override fun onFailure(reasonCode: Int) {
+        Log.d(TAG, "Disconnect failed. Reason: $reasonCode")
+        closeChannel()
+      }
+
+      override fun onSuccess() {
+        Log.d(TAG, "Disconnect successful")
+        closeChannel()
+      }
+    })
+  }
+
+  private fun closeChannel() {
+    if (VERSION.SDK_INT >= VERSION_CODES.O_MR1) {
+      channel.close()
+    }
+  }
+
+  fun getErrorMessage(reason: Int): String {
+    return when (reason) {
+      WifiP2pManager.ERROR -> "Internal error"
+      WifiP2pManager.BUSY -> "Framework busy, unable to service request"
+      WifiP2pManager.P2P_UNSUPPORTED -> "P2P unsupported on this device"
+      else -> "Unknown error code - $reason"
+    }
+  }
+
+  fun displayToast(
+    stringResourceId: Int,
+    templateValue: String,
+    duration: Int
+  ) = activity.toast(activity.getString(stringResourceId, templateValue), duration)
+
+  fun displayToast(stringResourceId: Int, duration: Int) =
+    activity.toast(stringResourceId, duration)
+
+  fun onFileTransferAsyncTaskComplete(wereAllFilesTransferred: Boolean) {
+    if (wereAllFilesTransferred) {
+      displayToast(R.string.file_transfer_complete, Toast.LENGTH_LONG)
+    } else {
+      displayToast(R.string.error_during_transfer, Toast.LENGTH_LONG)
+    }
+    callbacks.onFileTransferComplete()
+  }
+
+  interface Callbacks {
+    fun onUserDeviceDetailsAvailable(userDevice: WifiP2pDevice?)
+    fun onConnectionToPeersLost()
+    fun updateListOfAvailablePeers(peers: WifiP2pDeviceList)
+    fun onFilesForTransferAvailable(filesForTransfer: ArrayList<FileItem>)
+    fun onFileStatusChanged(itemIndex: Int)
+    fun onFileTransferComplete()
+  }
+
+  companion object {
+    private const val TAG = "WifiDirectManager"
+    var FILE_TRANSFER_PORT = 8008
+    @Throws(IOException::class) fun copyToOutputStream(
+      inputStream: InputStream,
+      outputStream: OutputStream
+    ) {
+      inputStream.use { it.copyTo(outputStream, DEFAULT_BUFFER_SIZE) }
```

Comment:
you could also call `.use` on the output stream and you wouldn't have to call close